docs/glossary.rst

```diff
@@ -0,0 +1,274 @@
+        extend the work graph (such as in support of adaptive work flows).
+
+        A worker may be launched to perform a single task, to perform a sequence
+        of tasks, or to participate in an :term:`ensemble`.
+
+    CPI
+        .. todo:: Define CPI.
+
+    Context
+      Abstraction for the entity that maps work to a computing environment.
+      Instances may be long-lived and participate in owning/managing work and
+      data references.
+
+    Session
+      Abstraction for the entity representing work that is executing on resources
+      allocated by an instance of a Context implementation. The Session is the
+      scoped active state of a Context while computing resources are held.
+
+.. topic:: *Context* and *Session*
+
+    roughly map to terms like *Executor* and *Task* in some other frameworks.
+    Distinctions relate to the lifetime of the :term:`Context` instance, and the fact that
+    it owns both the work specification (including operation and data handles)
+    and the computing resources.
+    The :term:`Context` instance owns resources (on behalf of the client) that may
+    otherwise be owned directly by the client, and so its lifetime must span all
+    references to resources, operation handles, and data futures.
+
+.. glossary::
+
+    discovered task
+        A task that has become runnable, but was not already scheduled.
+        This term is intentionally vague as the requirements and constraints of
+        work management are explored. The primary usage refers to a task that
+        has been generated due to adaptations in the work flow. It may also
+        apply to tasks that may be scheduled opportunistically, or simply to
+        the change of state when a task's input dependencies have been met.
+
+.. _user classification:
+
+User classes and characteristics
+--------------------------------
+
+Users are assumed to be molecular science researchers using Python scripts to
+express and execute simulation and analysis work consisting of multiple
+simulation and analysis tasks, using software tools from multiple packages.
+
+Software tools are individually accessible as Python modules or as command line
+tools.
+
+Computational work may require multiple invocations (multiple HPC jobs) to complete.
+
+The following classes of user are not necessarily mutually exclusive.
+
+.. glossary::
+
+    basic user
+        A researcher writing a Python script to control standard software.
+
+    advanced user
+        A researcher who needs to integrate custom code into the scripted work.
+
+    pure Python user
+        All software necessary for the work is importable as Python modules.
+
+    mixed command line user
+        Some software is only accessible to Python by wrapping a command line driven tool.
+
+    compiled extension user
+        Some software necessary for the work requires compilation and/or installation
+        on the computing resource.
+
+    direct user
+    local user
```

Comment:
@mrshirts writes:
> I think the distinction between these as it came up on the discussion, as local being being run on a single computational resource that the driving program is launched on, versus being run on a XSEDE or other computational resource, that usually requires a SLURM script to interact with it. The distinction not being the python interpreter, but the computational resource. The distinction between the two should be resolved.

Reply:
Right. There may be more or different ways to slice it. What are the relevant distinctions in how work is launched in terms of development milestones and usability / features? Single-process versus multi-process is likely relevant. Also: a script launched within the target execution environment versus a script that dispatches work to another execution environment. This should become clearer as #23 proceeds.